### PR TITLE
[TOD][Agents] Helper Agents for simulation + tests

### PR DESCRIPTION
**Pull request by moyapchen**
Effectively does same thing as https://github.com/facebookresearch/ParlAI/pull/4176 but broken out to be easier to review.

This diff does 2 things:
1. Adds the helper agents for simulation (Goal + API Description agents necessary for simulation; agents for dumping data from a dataset) in `tod_agents.py`. 
2. Adds tests for these agents + the teachers in the previous diff in the stack. 

This is the same as https://github.com/facebookresearch/ParlAI/pull/4232 but I pressed the wrong button there.

### CODE REVIEW

File: parlai/core/tod/tod_test_utils/test_agents.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+def get_round_utts(episode_idx, max_rounds, filter_utts=None):
+    if max_rounds < 1:
+        return []
+    utts = [
+        [
+            f"USER: user_utt_{episode_idx}_0",
+            "APICALL: ",
+            "APIRESP: ",
+            f"SYSTEM: sys_utt_{episode_idx}_0",
```
Comment (skiingpacman) [inline parlai/core/tod/tod_test_utils/test_agents.py:88]: why not use the tod_core constants for USER, APICALL, etc?

**Review by skiingpacman — APPROVED** [review]
Bunch of comments that should help make things clearer. Other than that looks good. (Note I only skimmed the unit tests - so possibly missed stuff there.)

File: parlai/core/tod/tod_agents.py [hunk 2]
```diff
@@ -116,6 +117,447 @@ def _get_agent_type_suffix(self) -> str:
...
+        result = []
+        for raw in raw_episodes:
+            for call in self.filter_goals(raw.goal_calls_machine):
+                schema = {}
+                for cand in raw.api_schemas_machine:
+                    if (
+                        cand[tod.STANDARD_API_NAME_SLOT]
+                        == call[tod.STANDARD_API_NAME_SLOT]
+                    ):
+                        schema = cand
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:257]: Nope, for k dialogue goals in a dialogue, there are now k dialogues generated each with a single goal.. 

This code says that if there are multiple schemas listed with the same API call name (which ideally shouldn't happen though I don't have validation for this active), it'll grab the last of those to use as the schema.

File: parlai/core/tod/tod_agents.py [hunk 3]
```diff
@@ -116,6 +117,447 @@ def _get_agent_type_suffix(self) -> str:
...
+    """
+
+    def _get_agent_type_suffix(self):
+        return "SingleApiSchema"
+
+
+###### Agents used for calculating TOD World Metrics based on a dataset. See `tod_world_script` or `parlai/projects/tod_simulator/` for examples.
+class TodUserUttAgent(_TodDataDumpAgent):
+    """
+    Agent used to calculate TOD World Metrics on a dataset. Represents the "User" agent.
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:303]: Yeah... Updated the comment to be a bit more verbose.

```
    Agent provided as a convenience to run TOD World script code on a dataset without having to write too much code to do so. (Ex. for a quick way to dump data to a `.jsonl` file for generating data for ACUTE or to generate a report file of metrics from TodWorld script.)
    ```

File: parlai/core/tod/tod_agents.py [hunk 4]
```diff
@@ -116,6 +117,447 @@ def _get_agent_type_suffix(self) -> str:
...
+    """
+    Agent used to calculate TOD World Metrics on a dataset. Represents the "System"
+    agent.
+
+    This class should only ever be used with the model-model chat world which will stop
+    upon seeing the '[DONE]' utterance; may go out of bounds otherwise.
+    """
+
+    def __init__(self, opt: Opt, shared=None):
+        # This class represents two "agents" so need to make sure we don't increment episode number (reset) twice
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:340]: Updated comment here to say. 

```
        # This class will have `act()` called on it twice per round — once for API call and once for NLG — so need to make sure we don't increment episode number (reset) prematurely; use the `already_reset` flag for this. ```

File: parlai/core/tod/tod_agents.py [hunk 5]
```diff
@@ -116,6 +117,447 @@ def _get_agent_type_suffix(self) -> str:
...
+        super().__init__(opt, shared)
+
+    def act(self):
+        self.already_reset = False
+        if tod.STANDARD_API_SCHEMAS in self.observation.get("text", ""):
+            return {
+                "text": tod.STANDARD_API_SCHEMAS,
+                "id": self.id,
+                "domain": self.episode.domain,
+                "episode_down": False,
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:352]: Well that's a weird typo lol.

File: parlai/core/tod/tod_agents.py [hunk 6]
```diff
@@ -116,6 +117,447 @@ def _get_agent_type_suffix(self) -> str:
...
+        # This class represents two "agents" so need to make sure we don't increment episode number (reset) twice
+        self.already_reset = False
+        self.api_call_turn = True
+        super().__init__(opt, shared)
+
+    def act(self):
+        self.already_reset = False
+        if tod.STANDARD_API_SCHEMAS in self.observation.get("text", ""):
+            return {
+                "text": tod.STANDARD_API_SCHEMAS,
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:349]: Ahhhhh this is a convention I put in for the first turn. Can leave it as a comment.

(At one point, when it wasn't clear exactly what the structure of the API response setup would be — and if we say, wanted to change the API Response based on if we were using structured or unstructured prompts — I made it such that the API Schema system would pass this information to the API Response agent as a preemptive measure. It's not a big enough of a thing to remove, but I can leave a comment to make it clear what's happening here.)

File: parlai/core/tod/tod_test_utils/test_agents.py (hunk 1 above)
Comment (moyapchen) [reply parlai/core/tod/tod_test_utils/test_agents.py:88]: Oof good minor catch.

File: parlai/core/tod/tod_agents.py [hunk 7]
```diff
@@ -116,6 +117,447 @@ def _get_agent_type_suffix(self) -> str:
...
+        else:
+            result = {
+                "text": f"{tod.STANDARD_SYSTEM_UTTERANCE}{self.episode.rounds[self.round_idx].sys_utt}",
+                "id": self.id,
+                "domain": self.episode.domain,
+                "episode_done": False,
+            }
+            self.round_idx += 1
+
+        self.api_call_turn ^= True
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:371]: I think I was being paranoid that I would forget to set that variable to the opposite of itself at one point.. will address.

**Pull request merged**